### PR TITLE
Add ? and * wildcards to filtered-search, closes #1006.

### LINKED ISSUE #1006: Support wildcards in filtered list for Subscriber Later Binding SMV and GOOSE
**Is your feature request related to a problem? Please describe.**

The filtered lists in OpenSCD are powerful and an integral part of the workflow.
However, they do not always facilitate an efficient workflow for particular device semantics.

Consider the following example:

image

* <1> and <2> show the internal addresses which are semantically in this relay (a SEL-400 series relay)
* <3> shows a second attachment point which is used to aggregate SV streams internally within the device.

The nomenclature is:

* `I` = Current
* `A` or `B` or `C` = Phase
* `S` or `T` or `U` = Terminal
* `_2` or `_3` is the attachment point to allow multiple SMV streams to be combined.

As an end-user I wish to select the terminal first (the terminal might be, for instance, a side of a transformer).
Because of these particular device semantics, this is at the end of the internal address, making this hard to do.
Ideally, our filtering would be powerful enough that this would not matter.

I want the following order of operations:

* `IAS`
* `IBS`
* `ICS`
* `IAS_2`
* `IBS_2`
* `ICS_2`

**Describe the solution you'd like**

I want to be able to filter: `I?S` and then `I?S_2` to improve my workflow.
This saves me from having to re-apply a filter for each phase.
At the moment, I think I only need a per-character wildcard.

**Test Files**

See attached:
[XAT_Prot1.scd.zip](https://github.com/openscd/open-scd/files/9587011/XAT_Prot1.scd.zip)

### PR DESCRIPTION
**Pull request by danyill**
This is just a prototype.

It does allow you to do a nice thing, i.e. line up phases between left and right.

image

### CODE REVIEW

**Comment by ca-d** [comment]
@danyill I was really intrigued when you just mentioned the possibility of letting users filter by regex. Would a "regex toggle button" next to the search field be a viable alternative for this use case as well?

**Comment by danyill** [comment]
> @danyill I was really intrigued when you just mentioned the possibility of letting users filter by regex. Would a "regex toggle button" next to the search field be a viable alternative for this use case as well?

Yes, that is a good idea. I am tempted to say let's just clone the Developer Tools from Chrome.

image

But I wonder if regex search shouldn't be a settings toggle or a "pro mode" feature, and more basic filtering like what we have now with just the addition of wildcards might be better for most users...

**Comment by JakobVogelsang** [comment]
I would propose to start with the wildcard. The implementation looks good to me. But tests are missing ;)

**Comment by danyill** [comment]
In fact, the implementation was not good, alas, but I hope it is better now. Please check carefully, we worry terribly about unintended consequences and the high impact of this change!

File: src/filtered-list.ts [hunk 1]
```diff
@@ -38,11 +39,23 @@ function hideFiltered(item: ListItemBase, searchText: string): void {
...
-  const terms: string[] = searchText.toUpperCase().split(' ');
-
-  terms.some(term => !filterTarget.includes(term))
-    ? slotItem(item).classList.add('hidden')
-    : slotItem(item).classList.remove('hidden');
+  const terms: string[] = searchText
+    .toUpperCase()
+    .replace(/[.+^${}()|[\]\\]/g, '\\$&')
+    .trim()
+    .split(' ');
```
Comment (JakobVogelsang) [inline src/filtered-list.ts:46]: split on `/\s/g`

File: src/filtered-list.ts [hunk 2]
```diff
@@ -38,11 +39,23 @@ function hideFiltered(item: ListItemBase, searchText: string): void {
...
   ).toUpperCase();
 
-  const terms: string[] = searchText.toUpperCase().split(' ');
-
-  terms.some(term => !filterTarget.includes(term))
-    ? slotItem(item).classList.add('hidden')
-    : slotItem(item).classList.remove('hidden');
+  const terms: string[] = searchText
+    .toUpperCase()
+    .replace(/[.+^${}()|[\]\\]/g, '\\$&')
```
Comment (JakobVogelsang) [inline src/filtered-list.ts:44]: Where did you get the collection of "problematic" characters from?

**Review by JakobVogelsang — DISMISSED** [review]
One little change and one question other that that VERY WELL DONE!

File: src/filtered-list.ts (hunk 2 above)
Comment (danyill) [reply src/filtered-list.ts:44]: I expect them somewhere on SO, but I couldn't find them again for looking. They look like the right ones for a regex on the face of it.

File: src/filtered-list.ts (hunk 1 above)
Comment (JakobVogelsang) [reply src/filtered-list.ts]: I know thank you :)

**Comment by danyill** [comment]
Thank you for this review. I have made a little change to use a regex for multiple spaces for the split.

**Review by JakobVogelsang — APPROVED** [review]
LGTM

**Pull request merged**